File: 动态规划/chapter10/No902/main.py
from functools import cache
from typing import List


class Solution:
    def atMostNGivenDigitSet(self, digits: List[str], n: int) -> int:
        s = str(n)

        @cache
        def dfs(i: int, is_limit: bool, is_num: bool) -> int:
            if i == len(s):
                return int(is_num)
            res = 0
            if not is_num:
                res = dfs(i + 1, False, False)
            up = int(s[i]) if is_limit else 9
            for x in digits:
                x = int(x)
                if x > up:
                    break
                res += dfs(i + 1, is_limit and x == up, True)

            return res

        return dfs(0, True, False)


def main():
    digits = ["1", "4", "9"]
    n = 1000000000
    sol = Solution()
    print(sol.atMostNGivenDigitSet(digits, n))


main()

File: 动态规划/chapter10/No902/main.cpp
#include <functional>
#include <iostream>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
  int atMostNGivenDigitSet(vector<string> &digits, int n) {
    string s = to_string(n);
    int m = s.length();
    vector<int> f(m, -1);

    function<int(int, bool, bool)> dfs = [&](int i, bool isLimit,
                                             bool isNum) -> int {
      if (i == m)
        return isNum;
      if (!isLimit && isNum && f[i] >= 0) {
        return f[i];
      }
      int res = 0;
      if (!isNum) {
        res = dfs(i + 1, false, false);
      }
      char up = isLimit ? s[i] : '9';
      for (auto &d : digits) {
        if (d[0] > up)
          break;
        res += dfs(i + 1, isLimit && d[0] == up, true);
      }
      if (!isLimit && isNum) {
        f[i] = res;
      }
      return res;
    };

    return dfs(0, true, false);
  }
};

signed main() {
  vector<string> digits = {"1", "4", "9"};
  int n = 1000000000;
  Solution sol;
  cout << sol.atMostNGivenDigitSet(digits, n) << endl;

  return 0;
}